posts/2020/08/reproducible-builds.md

```diff
@@ -0,0 +1,146 @@
+as deterministic things, they often are not, and on purpose. Mainly driven by security concern, a number of processes in
+a computer use pseudo-randomness to prevent an attacker from gaining control over a system. Thus by default, many of the
+typical actions you may do in a software (iterating over a dictionary in Python), with have _some_ randomness in them,
+which impact the determinism of a process.
+
+There are of course a number of source of involuntary randomness doe to the practicality of computer systems. Among
+other, the current date and time, your user name and uid, hostname, order of files on your hard drive, and in which
+order they may be listed...
+
+To obtain a reproducible results, one thus often need to make sure each step is deterministic, and that all the variable
+influencing that process are either controlled, or recorded.
+
+# Reproducible artifact build
+
+In IPython 7.16.1 we have tried to removed all sources of variability, by controlling the order in which the files are
+generated, archived in the sdists, their metadata, timestamp... etc. Thus should be able to go from the commit in the
+source repository to the sdist/wheel and obtain an identical result. This should help you to trust that no backdoor has
+been introduced in the released packages, though this is also critically useful for efficiency in package managers.
+
+Of course you have to trust that the IPython source itself, and its dependencies are devoid of backdoors, but let's move
+one step at a time. Reproducible build artifacts can also have impact on the build and installation process of packages.
+
+# Efficient dependencies rebuild. 
+
+Currently IPython depends on many packages: prompt_toolkit, traitlets, setuptools, ...etc, and we have a number of
+dependees, ipykernel, then jupyter notebook... When dependencies tree is rebuilt for a reason or another, a change of a
+single bit could trigger the rebuild of the all chain. When package like IPython are not reproducible, this mean  a
+rebuild of IPython – whether it has changed or not – could trigger a rebuild of all downstream elements.
+
+With reproducible build, you can trust that the artifact will not change after a rebuild. For the functional programmer
+around you it indicate that the process of building from the source is a pure function. Therefore it can safely be part
+of a distributed system (rebuilding on two different places will give the same result, so you can avoid costly data
+movement), and we can also do caching of results, so for identical input we know the output will be identical.
+
+This can allow break rebuild chains and stop as soon as soon as a dependency rebuild has no effect.
+
+This has the opportunity to massively decrease the time spend by platform like conda-forge on rebuilding the ecosystem
+on new version of Python; making new packages available faster. 
```

Comment:
It sounds reasonable, but I'm sure if this is actually true. E.g. wouldn't the `.pyc` files change between Python versions? Or are those not present in a conda package?

@scopatz does conda-forge benefit here?

Reply:
I not certain that conda-forge would benefit here because IPython seems to have different dependencies on different platforms (like colorama on windows). So each build artifact is different in a meaningful way. These differences would need to be removed for a noarch package to be made.

Reply:
I am all for moving in this direction, though!

Reply:
BTW I think this is really hard for a cross-platform shell.  xonsh has the exactly the same issue.

Reply:
That's where I think we could move to conda-forge having metadata and data actually separate, there are several other factors for that, like for example dependencies being wrong and needing a rebuild or a metadata hotfix.

There are other things that change on a per-python version; like the python path includes the python version. Though having reproducible build is necessary to have identical artifacts across version; but of course not sufficient. Let me know if you want me to remove, or expand this section.

Reply:
I think I'll leave that up to Ralf.  But yes, I agree.  We need a better way in conda (not the forge per se) to declare noarch deps and metadata

Reply:
Maybe saying explicitly "reproducible builds are a necessary but not sufficient condition ..." and keeping the statement would be good.

Reply:
how about this I just pushed?

```diff
This can allow break rebuild chains and stop as soon as soon as a dependency rebuild has no effect.

-This has the opportunity to massively decrease the time spend by platform like conda-forge on rebuilding the ecosystem
+Thus, reproducible builds are a necessary but not sufficient condition to decrease the time spend by platform like conda-forge on rebuilding the ecosystem
 on new version of Python; making new packages available faster.
```

Reply:
That SGTM

Reply:
:+1: